File: code/Kotlin/libs/annotation_processing/apt/compiler/src/main/java/inject/processor/creator/FragmentMethod.java
package inject.processor.creator;

import static javax.lang.model.element.Modifier.PUBLIC;

import com.squareup.javapoet.MethodSpec;
import com.squareup.javapoet.TypeName;

import javax.lang.model.element.Element;

import inject.classname.AndroidClassName;
import inject.classname.OtherClassName;
import inject.classname.PrimitiveClassName;

/**
 * @author yuansui
 * @since 2020/9/8
 */
class FragmentMethod extends BaseMethod {

    public FragmentMethod(CreatorProcessor processor, Element annotatedElement) {
        super(processor, annotatedElement);
    }

    MethodSpec get() {
        MethodSpec.Builder builder = MethodSpec.methodBuilder("get")
                .addModifiers(PUBLIC)
                .addStatement("$T b = new $T()", AndroidClassName.BUNDLE, AndroidClassName.BUNDLE);

        addBundleStatement(builder);

        builder.addStatement("$T frag = new $T()", annotatedTypeName, annotatedTypeName)
                .addStatement("frag.setArguments(b)")
                .addStatement("return frag")
                .returns(annotatedTypeName);

        return builder.build();
    }

    private void addBundleStatement(MethodSpec.Builder builder) {
        for (Element e : allFields) {
            String paramName = processor.getParamName(e);
            builder.beginControlFlow("if ($N != null)", paramName);
            TypeName typeName = processor.getTypeNameBox(e);
            if (typeName.isBoxedPrimitive()) {
                // Long Boolean Integer...
                if (typeName.unbox() == TypeName.INT) {
                    builder.addStatement("b.put$N($S, $N)", "Int", paramName, paramName);
                } else {
                    builder.addStatement("b.put$T($S, $N)", typeName, paramName, paramName);
                }
            } else {
                if (typeName.equals(PrimitiveClassName.STRING)) {
                    builder.addStatement("b.putString($S, $N)", paramName, paramName);
                } else {
                    processor.addEson(builder);
                    builder.beginControlFlow("try");
                    builder.addStatement("b.putString($S, eson.toJson($N))", paramName, paramName);
                    builder.nextControlFlow("catch($T e)", OtherClassName.EXCEPTION);
                    builder.addStatement("e.printStackTrace()");
                    builder.endControlFlow();
                }
            }

            builder.endControlFlow();
        }
    }
}

File: code/Kotlin/libs/annotation_processing/kapt/compiler/src/main/java/inject/compiler/method/creator/FragmentMethod.kt
package inject.compiler.method.creator

import com.squareup.kotlinpoet.FunSpec
import inject.compiler.classname.AndroidClassName
import inject.compiler.classname.OtherClassName
import inject.compiler.method.creator.BaseMethod
import inject.compiler.ext.addEson
import inject.compiler.ext.asTypeName
import inject.compiler.ext.getParamName
import inject.compiler.ext.isParcelable
import inject.compiler.ext.isPrimitive
import inject.compiler.ext.isSerializable
import inject.compiler.ext.isString
import inject.compiler.processor.FormatSpecifiers
import inject.compiler.processor.CreatorProcessor
import javax.lang.model.element.Element

/**
 * @author yuansui
 * @since 2020/9/8
 */
internal class FragmentMethod(
    processor: CreatorProcessor,
    annotatedElement: Element
) : BaseMethod<CreatorProcessor>(processor, annotatedElement) {

    fun get(): FunSpec {
        val builder: FunSpec.Builder = FunSpec.builder("get")
            .addStatement("val b = ${FormatSpecifiers.TYPE}()", AndroidClassName.BUNDLE)

        addBundleStatement(builder)

        builder.addStatement("val frag = ${FormatSpecifiers.TYPE}()", annotatedTypeName)
            .addStatement("frag.setArguments(b)")
            .addStatement("return frag")
            .returns(annotatedTypeName)

        return builder.build()
    }

    private fun addBundleStatement(builder: FunSpec.Builder) {
        builder.addEson()
        for (e in allFields) {
            val paramName = e.getParamName()
            val typeName = e.asTypeName()
            if (typeName.isPrimitive() || typeName.isString() || typeName.isSerializable() || typeName.isParcelable()) {
                builder.addStatement(
                    "${FormatSpecifiers.NAME}?.let { b.put${FormatSpecifiers.TYPE}(${FormatSpecifiers.STRING}, it) }",
                    paramName,
                    typeName,
                    paramName
                )
            } else {
                builder.beginControlFlow("try")
                builder.addStatement(
                    "b.putString(${FormatSpecifiers.STRING}, eson.toJson(${FormatSpecifiers.NAME}))",
                    paramName,
                    paramName
                )
                builder.nextControlFlow("catch(e: ${FormatSpecifiers.TYPE})", OtherClassName.EXCEPTION)
                builder.addStatement("e.printStackTrace()")
                builder.endControlFlow()
            }
        }
    }
}
